lib/spack/spack/util/environment.py

```diff
@@ -446,7 +446,12 @@ def from_sourcing_file(filename, *args, **kwargs):
                              for k, v in env_after.items())
 
         # Other variables unrelated to sourcing a file
-        blacklist.extend(['SHLVL', '_', 'PWD', 'OLDPWD', 'PS2'])
+        blacklist.extend([
+            # Bash internals
+            'SHLVL', '_', 'PWD', 'OLDPWD', 'PS2', 'ENV',
+            # Environment modules v4
+            'LOADEDMODULES', '_LMFILES_', 'BASH_FUNC_module()'
```

Comment:
If you want to blacklist all Modules runtime variables, you should also add:
* `MODULES_LMALTNAME`
* `MODULES_LMCONFLICT`
* `MODULES_LMNOTUASKED`
* `MODULES_LMPREREQ`

And for the [item reference counting feature](https://modules.readthedocs.io/en/latest/MIGRATING.html#path-variable-element-counter) in path-like variable:
* `<VAR>_modshare`
* `MODULES_MODSHARE_<VAR>` (used instead of the above for the DYLD-related variables)

For the record, all Modules-related environment variable can be found at https://modules.readthedocs.io/en/latest/module.html#environment

Reply:
This has been taken care of in:

https://github.com/spack/spack/blob/9345436ff8c1bf18008b3a59eeb0473abc074a20/lib/spack/spack/util/environment.py#L419-L421

and is tested in:

https://github.com/spack/spack/blob/9345436ff8c1bf18008b3a59eeb0473abc074a20/lib/spack/spack/test/environment_modifications.py#L367-L372